data/src/main/java/com/microsoft/azure/kusto/data/auth/CloudDependantTokenProviderBase.java

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.kusto.data.auth;
+
+import com.microsoft.azure.kusto.data.exceptions.DataClientException;
+import com.microsoft.azure.kusto.data.exceptions.DataServiceException;
+import java.net.URISyntaxException;
+import java.util.HashSet;
+import java.util.Set;
+import org.jetbrains.annotations.NotNull;
+
+public abstract class CloudDependantTokenProviderBase extends TokenProviderBase {
+    private static final String ERROR_INVALID_SERVICE_RESOURCE_URL =
+            "Error determining scope due to invalid Kusto Service Resource URL";
+    protected final Set<String> scopes = new HashSet<>();
+    protected CloudInfo cloudInfo = null;
+
+    CloudDependantTokenProviderBase(@NotNull String clusterUrl) throws URISyntaxException {
+        super(clusterUrl);
+    }
+
+    protected void onCloudInfoInitialized() throws DataClientException, DataServiceException {
+        try {
+            scopes.add(cloudInfo.determineScope());
+        } catch (URISyntaxException e) {
+            throw new DataServiceException(clusterUrl, ERROR_INVALID_SERVICE_RESOURCE_URL, e, true);
+        }
+    }
+
+    protected void initializeCloudInfo() throws DataClientException, DataServiceException {
```

Comment:
I might be confusing things here, but I thought we talked about encapsulating any `CloudInfo` code into that class.
Do we need to maintain a `CloudInstance` here, as opposed to that being part of `CloudInfo`'s `Map` and we just retrieve what we need by indexing with our `clusterUrl`?

Reply:
Cloudinfo is no longer a member, let me know if that's what you had in mind?

Reply:
I don't recall the different approaches I considered, but this is a very good one. Most importantly, `CloudInfo` is not longer a member of these classes, which makes a lot more sense.